Enable per channel zero point.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37829 Some TODO fixes.
* #37626 Perf optimization for conv and gemm kernels.
* #37624 Added per channel separate test cases for fc and deconv tests.
* #37623 Changes to enable per channel support on dynamic linear.
* #37622 Enabled per channel quantized static linear/conv
* #37621 Added per channel kernels for depthwise conv.
* #37620 Changes to enable per channel requant.
* **#37619 Enable per channel zero point.**
* #37618 Interface changes to enable per channel quant.

Summary:
This PR introduces changes to add per channel zero point. Now the kernels use output_channel_index that is passed to the kernel. This is used to lookup the corresponding zero point.
Requantization scale does not use output_channel_index yet. This is done in the next diff.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

kimishpatel committed:
Enable per channel zero point.

Summary:
This PR introduces changes to add per channel zero point.
Modifies kernels appripriately.
Some bug fixes in enabling per channel zero point.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point.
Modifies kernels appripriately.
Some bug fixes in enabling per channel zero point.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point.
Modifies kernels appripriately.
Some bug fixes in enabling per channel zero point.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point.
Modifies kernels appripriately.
Some bug fixes in enabling per channel zero point.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point.
Modifies kernels appripriately.
Some bug fixes in enabling per channel zero point.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point.
Modifies kernels appripriately.
Some bug fixes in enabling per channel zero point.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point.
Modifies kernels appripriately.
Some bug fixes in enabling per channel zero point.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point.
Modifies kernels appripriately.
Some bug fixes in enabling per channel zero point.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point. Now the kernels use output_channel_index that is passed to the kernel. This is used to lookup the corresponding zero point.
Requantization scale does not use output_channel_index yet. This is done in the next diff.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point. Now the kernels use output_channel_index that is passed to the kernel. This is used to lookup the corresponding zero point.
Requantization scale does not use output_channel_index yet. This is done in the next diff.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/convolution.c:
```diff
@@ -198,16 +175,14 @@ enum pytorch_qnnp_status pytorch_qnnp_create_convolution2d_nhwc_q8(
+    if (requantization_scales[i] <= 0.0f ||
+        !isnormal(requantization_scales[i])) {
+      pytorch_qnnp_log_error(
+          "failed to create fully connected operator with %.7g requantization scale: scale must be finite and positive",
```
nit: update message

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/deconvolution.c:
```diff
@@ -103,39 +101,17 @@ enum pytorch_qnnp_status pytorch_qnnp_create_deconvolution2d_nhwc_q8(
+    if (requantization_scales[i] <= 0.0f ||
+        !isnormal(requantization_scales[i])) {
+      pytorch_qnnp_log_error(
+          "failed to create convolution operator with %.7g requantization scale for "
```
message should say deconvolution

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point. Now the kernels use output_channel_index that is passed to the kernel. This is used to lookup the corresponding zero point.
Requantization scale does not use output_channel_index yet. This is done in the next diff.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point. Now the kernels use output_channel_index that is passed to the kernel. This is used to lookup the corresponding zero point.
Requantization scale does not use output_channel_index yet. This is done in the next diff.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

dreiss reviewed (approved):
Please update the messages that Supriya pointed out.  I didn't review the microkernels, but I trust your tests.  Ashkan, feel free to review them.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8conv/8x8-aarch64-neon.S:
```diff
@@ -19,10 +37,11 @@
 #    uint8_t* restrict c,
 #    size_t c_stride,
+#    size_t output_channel_index,
 #    const union pytorch_qnnp_q31_requantization_params quantization_params[restrict static 1])
```
It seems to me this should be pytorch_qnnp_conv_quantization_params not pytorch_qnnp_q31_requantization_params according to q8conv.h

kimishpatel replied:
Yes. This was a typo from original one. Gonna skip fixing in these PRs.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8conv/8x8-aarch64-neon.S:
```diff
@@ -31,17 +50,27 @@ BEGIN_FUNCTION pytorch_q8conv_ukernel_8x8__aarch64_neon
+    # We do not need to multiply with size of pointer type
+    LDR x9, [sp]
+    # Load pointer to per channel zero points array
+    LDR x10, [x8]
```
How about replacing this with `LDR x10, [x8], 8` and getting rid of `ADD x8, x8, 8` below.

kimishpatel replied:
Yeah tried that. Didn't compile. It is a valid instruction but does not compile on android.

AshkanAliabadi replied:
That's strange.  I did a quick grep on XNNPACK to see if they have any such usage and there is:

https://github.com/google/XNNPACK/blob/8bd4cbae92d8080272fcae3aa0509ad49375bbe0/src/f32-gemm/4x12-aarch64-neonfma-cortex-a53.S.in#L176

It's not a big deal.  Will barely make a difference in isolation but I was hoping we could avoid it.  At least this is not in the inner most loop - although it is in a loop - the one that goes over the tiles.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8conv/8x8-aarch64-neon.S:
```diff
@@ -31,17 +50,27 @@ BEGIN_FUNCTION pytorch_q8conv_ukernel_8x8__aarch64_neon
     # Load a_zero_point
-    LD1R {v24.8b}, [x8], x9
+    LD1R {v24.8b}, [x8]
+    ADD x8, x8, 4
```
The two can be replaced with `LD1R {v24.8b}, [x8], 4`.

kimishpatel replied:
Same as above.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/8x8-aarch64-neon.S:
```diff
@@ -9,6 +9,23 @@
 #include <qnnpack/assembly.h>
 #include <requantization/runtime-assembly.h>
 
```
Same comments as 8x8-aarch64-neon.S apply.

kimishpatel replied:
Most likely same responses as well :).

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/8x8-aarch64-neon.S:
```diff
@@ -30,15 +48,23 @@ BEGIN_FUNCTION pytorch_q8gemm_ukernel_8x8__aarch64_neon
+    LDR x8, [sp, 8]
+
+    # Load output channel index
+    LDR x16, [sp]
```
`LDP x16, x8, [sp]`, etc.  Same for the rest.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8conv/4x8-aarch32-neon.S:
```diff
@@ -56,10 +95,12 @@ BEGIN_FUNCTION pytorch_q8conv_ukernel_4x8__aarch32_neon
     # - d14 = a_zero_point
-    VLD1.8 {d14[]}, [r9], r4
+    VLD1.8 {d14[]}, [r9]
+    ADD r9, r9, 4
```
If you have a scratch register to spare you can fold the post increment into VLD1.

kimishpatel replied:
I think there was only one instance of it here. To use r4 I would have to `move 4, #4`. So it was a wash. Earlier it made a difference because there was one more place that did VLD1 so you could save ADD there. Although if you end up having a dependency on r9 then I am not sure what happens. VLD1 has 3/4 cycle latency and if we are doing post-indexing then do you see 3/4 cycles latency on r9 as well? Not clear to me and it depends on when it is incremented and whether there is a bypass path for it or not. I dont think it adds or subtract much from performance anyway.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/convolution.c:
```diff
@@ -264,7 +239,7 @@ enum pytorch_qnnp_status pytorch_qnnp_create_convolution2d_nhwc_q8(
 #if !PYTORCH_QNNPACK_RUNTIME_QUANTIZATION
               input_zero_point,
-              kernel_zero_point,
+              kernel_zero_points[0],
```
Yes, correct. This code path is not exercised btw, ever. I am gonna cleanup in next few weeks. Its too much burden to carry.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8conv/4x4c2-sse2.c:
```diff
@@ -30,8 +31,28 @@ void pytorch_q8conv_ukernel_4x4c2__sse2(
-  const __m128i vb_zero_point = _mm_load_si128(
-      (const __m128i*)quantization_params->sse2.kernel_zero_point);
+  const int16_t vb_zero_point_0 =
+    (int16_t)(uint16_t)quantization_params->sse2.kernel_zero_points[
```
Ok will remove this one.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/params.h:
```diff
@@ -128,7 +128,7 @@ union pytorch_qnnp_q31_requantization_params {
 union pytorch_qnnp_conv_quantization_params {
   struct {
-    int32_t kernel_zero_point;
+    const uint8_t* kernel_zero_points;
```
Yes that is correct.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/conv-run.cc:
```diff
@@ -95,6 +95,7 @@ static void compute_q8gemm(
   uint8_t* c = context->c;
   const size_t c_stride = context->c_stride;
 
+  size_t output_channel_index = nr_block_start + group_index * n;
```
LOL.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/8x8-neon.c:
```diff
@@ -74,7 +75,8 @@ void pytorch_q8gemm_ukernel_8x8__neon(
   const uint8x8_t vb_zero_point =
-      vld1_dup_u8((const uint8_t*)&quantization_params->neon.kernel_zero_point);
+      vld1_u8((const uint8_t*)&quantization_params->neon.kernel_zero_points
+          [output_channel_index]);
```
Yes, could be written that way as well. Not sure if it makes a difference though. Maybe fewer keystrokes.

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point. Now the kernels use output_channel_index that is passed to the kernel. This is used to lookup the corresponding zero point.
Requantization scale does not use output_channel_index yet. This is done in the next diff.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

AshkanAliabadi reviewed (commented):
LGTM.  Thanks.

kimishpatel commented:
It is strange. Not sure why it did not let me do that before, but seems to work now. Let me make that change.

kimishpatel commented:
I will make the change as part of this PR: #37626. Will be easier to deal with rebase issues also.

kimishpatel commented:
> That's strange. I did a quick grep on XNNPACK to see if they have any such usage and there is:

> https://github.com/google/XNNPACK/blob/8bd4cbae92d8080272fcae3aa0509ad49375bbe0/src/f32-gemm/4x12-aarch64-neonfma-cortex-a53.S.in#L176

> It's not a big deal. Will barely make a difference in isolation but I was hoping we could avoid it. At least this is not in the inner most loop - although it is in a loop - the one that goes over the tiles.

Aah. It was LD1R and not LDR.

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point. Now the kernels use output_channel_index that is passed to the kernel. This is used to lookup the corresponding zero point.
Requantization scale does not use output_channel_index yet. This is done in the next diff.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point. Now the kernels use output_channel_index that is passed to the kernel. This is used to lookup the corresponding zero point.
Requantization scale does not use output_channel_index yet. This is done in the next diff.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point. Now the kernels use output_channel_index that is passed to the kernel. This is used to lookup the corresponding zero point.
Requantization scale does not use output_channel_index yet. This is done in the next diff.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point. Now the kernels use output_channel_index that is passed to the kernel. This is used to lookup the corresponding zero point.
Requantization scale does not use output_channel_index yet. This is done in the next diff.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point. Now the kernels use output_channel_index that is passed to the kernel. This is used to lookup the corresponding zero point.
Requantization scale does not use output_channel_index yet. This is done in the next diff.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]

kimishpatel committed:
Update on "Enable per channel zero point."


Summary:
This PR introduces changes to add per channel zero point. Now the kernels use output_channel_index that is passed to the kernel. This is used to lookup the corresponding zero point.
Requantization scale does not use output_channel_index yet. This is done in the next diff.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339044](https://our.internmc.facebook.com/intern/diff/D21339044)

[ghstack-poisoned]
